src/Type/Constant/ConstantArrayType.php

```diff
@@ -839,14 +839,29 @@ public function popArray(): Type
 
 	private function reverseConstantArray(TrinaryLogic $preserveKeys): self
 	{
-		$keyTypesReversed = array_reverse($this->keyTypes, true);
-		$keyTypes = array_values($keyTypesReversed);
-		$keyTypesReversedKeys = array_keys($keyTypesReversed);
-		$optionalKeys = array_map(static fn (int $optionalKey): int => $keyTypesReversedKeys[$optionalKey], $this->optionalKeys);
+		if ($preserveKeys->yes()) {
+			$keyTypesReversed = array_reverse($this->keyTypes, true);
+			$keyTypes = array_values($keyTypesReversed);
+			$keyTypesReversedKeys = array_keys($keyTypesReversed);
+			$optionalKeys = array_map(static fn (int $optionalKey): int => $keyTypesReversedKeys[$optionalKey], $this->optionalKeys);
 
-		$reversed = new self($keyTypes, array_reverse($this->valueTypes), $this->nextAutoIndexes, $optionalKeys, TrinaryLogic::createNo());
+			return new self($keyTypes, array_reverse($this->valueTypes), $this->nextAutoIndexes, $optionalKeys, TrinaryLogic::createNo());
+		}
+
+		$builder = ConstantArrayTypeBuilder::createEmpty();
+		for ($i = count($this->keyTypes) - 1; $i >= 0; $i--) {
+			$offsetType = $this->keyTypes[$i]->isInteger()->no()
+				? $this->keyTypes[$i]
+				: null;
+			$builder->setOffsetValueType($offsetType, $this->valueTypes[$i], $this->isOptionalKey($i));
+		}
 
-		return $preserveKeys->yes() ? $reversed : $reversed->reindex();
+		$array = $builder->getArray();
+		if (!$array instanceof self) {
+			throw new ShouldNotHappenException();
```

Comment:
I think we should remove this condition and just change the return type of this method to `Type`. It's private so no big deal.

Reply:
might be able to improve this and deprecate old methods by creating `Type::sliceArray(..): Type` and `Type::chunkArray(..): Type`, but not sure if that's needed / worth it.

Reply:
I'd be rather more safe and I propose these steps:

1) Remove the ShouldNotHappenException
2) At the place where reverseConstantArray is called, check if the returned type is ConstantArrayType and only after that call `slice`...

Reply:
But otherwise Type::sliceArray and Type::chunkArray would be nice :)

Reply:
> I'd be rather more safe and I propose these steps:
> 
>     1. Remove the ShouldNotHappenException
> 
>     2. At the place where reverseConstantArray is called, check if the returned type is ConstantArrayType and only after that call `slice`...

you mean without throwing exceptions, and e.g. returning the array right away if it's not a constant array anymore, right? would mean more instanceof I suppose, but I think I can clean those up later when adding new type methods

Reply:
but it feels a bit weird, it should really not happen I suppose xD

Reply:
going to do the refactor first..

Reply:
all good now :)